Add support for CPU scalar in addcmul

emmett-bicker opened the pull request:
Step required for performance in #143122 

Adds support for CPU scalar for tensor_2 in addcmul. For example:
```
import torch
a = torch.rand(2, 2, device="cuda")
b = torch.tensor(1e-3)

torch.add(a, b)
torch.addcmul(a, a, b)  # used to fail, now works
```

emmett-bicker committed:
Add support for CPU scalar in addcmul

emmett-bicker committed:
(c * b) -> (b* c) lol

janeyx99 commented on aten/src/ATen/native/PointwiseOps.cpp:
```diff
@@ -40,6 +48,7 @@ TORCH_META_FUNC(addcdiv)
         "The future addcdiv behavior is just the latter implementation: ",
         "(input + value * tensor1 / tensor2), for all dtypes.");
   }
+  // TODO: add support for cpu_scalar in addcdiv
```
We'd prefer tracking issues in the GitHub repo over TODOs in the codebase so feel free to convert this TODO into a feature request issue! But let's try to minimize the TODOs in the codebase.

EmmettBicker replied:
That makes sense! I'll do that now.

janeyx99 commented on aten/src/ATen/native/cuda/PointwiseOpsKernel.cu:
```diff
@@ -11,6 +11,12 @@
 
+void addcmul_cuda_scalar_tensor2_kernel(
+  TensorIteratorBase& iter,
+  const Scalar& tensor2_scalar,
```
haha bikeshedding, not sure if scalar2 is more concise/less confusing than tensor2_scalar....

janeyx99 replied:
yea i do think scalar_tensor2 is better

emmett-bicker committed:
Add conditional around string def

janeyx99 commented on aten/src/ATen/native/cuda/PointwiseOpsKernel.cu:
```diff
@@ -25,6 +31,11 @@ void addcmul_cuda_kernel(TensorIteratorBase& iter, const Scalar& value) {
         auto alpha = value.to<scalar_t>();
         static const auto addcmul_string = jiterator_stringify(
           template <typename T> T addcmul(T a, T b, T c, T alpha) { return a + alpha * (b * c); });
+        if (iter.is_cpu_scalar(3)) {
```
The code duplication is making me sad but I don't see a trivial way to detangle the AT_DISPATCH....

oh well maybe we don't worry about that this PR (as there was similar code duplication in lerp too)

EmmettBicker replied:
Do you think a function that takes in iter and the position of the cpu scalar and calls the scalar kernel would be better to reduce duplication?

EmmettBicker replied:
Oh I just realized why that wouldn't work since the return wouldn't properly exit the function without some boilerplate

janeyx99 replied:
Yea haha it's the scalar_t that the dispatch provides that we no longer have.

janeyx99 reviewed (commented):
Looks good to me overall, will approve on green CI. Left some nits. I think adding the error messaging would be nice but not necessary here.

emmett-bicker committed:
Added checks and removed TODO

janeyx99 commented on aten/src/ATen/native/cuda/PointwiseOpsKernel.cu:
```diff
@@ -21,6 +21,18 @@ void addcmul_cuda_scalar_tensor2_kernel(
+    "CPU Scalar support for tensor1 argument is not supported. "
+    "However, CPU Scalar support for tensor2 is supported, "
+    "please swap your tensor1 and tensor2 terms."
+  );
```
Can you add a test for these?

EmmettBicker replied:
Yes absolutely!

emmett-bicker committed:
Updated tests

janeyx99 commented on test/test_torch.py:
```diff
@@ -3319,6 +3319,30 @@ def rand_tensor(size, dtype, device):
+        if device == "cpu":
+            torch.addcmul(a, scalar, c, value=alpha)
+            torch.addcmul(scalar, b, c, value=alpha)
+        elif device == "cuda:0":
```
better to check that device type is cuda, or string starts with cuda as this test doesn't always run with device 0

EmmettBicker replied:
Ok! That makes a lot of sense

janeyx99 commented on test/test_torch.py:
```diff
@@ -3319,6 +3319,30 @@ def rand_tensor(size, dtype, device):
+        c = torch.rand((2, 2), device=device)
+        scalar = torch.rand([], device="cpu")
+
+        torch.addcmul(a, b, c, value=alpha)
+        torch.addcmul(a, b, scalar, value=alpha)  # supported on both
```
this should already be tested in the test above, yes? If so, then this doesn't have to be here. Or we can move the functionality from the above test here, and change the name of this test to not specify errors.

EmmettBicker replied:
Ok! I'll delete it from the previous function. I also feel like I should just not have called it errors since I test the cpu stuff so I'll change the name too!

janeyx99 commented on test/test_torch.py:
```diff
@@ -3319,6 +3319,30 @@ def rand_tensor(size, dtype, device):
+        torch.addcmul(a, b, scalar, value=alpha)  # supported on both
+
+        if device == "cpu":
+            torch.addcmul(a, scalar, c, value=alpha)
+            torch.addcmul(scalar, b, c, value=alpha)
```
It would be better to assert that these equal'd a reference impl vs just not erroring

janeyx99 commented on test/test_torch.py:
```diff
@@ -3319,6 +3319,30 @@ def rand_tensor(size, dtype, device):
             self.assertTrue(not (out.isnan() or out.isinf()))
 
+    def test_addcmul_cpu_scalar_errors(self, device):
+        # Logic is dtype agnostic, so dtype isn't tested
```
huh is this the way to not have dtype be parametrized? Could you look through the logs and check that this test isn't just running the same thing 8 times anyway?

EmmettBicker replied:
It only ran 2 times because I was doing "print(device)" while debugging -- I assumed it would work because the other tests had a @dtypes but not @devices

emmett-bicker committed:
Addressed comments (more in description)

name of function changed to increase clarity (test_addcmul_cpu_scalar_errors -> test_addcmul_cpu_scalars)

removed redundant check

cpu changed to assertEqual instead of nocrash

lint

janeyx99 commented on test/test_torch.py:
```diff
@@ -3315,6 +3319,28 @@ def rand_tensor(size, dtype, device):
+                             a + scalar * c * alpha)
+            self.assertEqual(torch.addcmul(a, scalar, c, value=alpha),
+                             a + scalar * c * alpha)
+        elif device[0:4] == "cuda":
```
```suggestion
        elif device.startswith("cuda"):
```

janeyx99 commented on test/test_torch.py:
```diff
@@ -3315,6 +3319,28 @@ def rand_tensor(size, dtype, device):
+        if device == "cpu":
+            self.assertEqual(torch.addcmul(a, scalar, c, value=alpha),
+                             a + scalar * c * alpha)
+            self.assertEqual(torch.addcmul(a, scalar, c, value=alpha),
```
this is the same as 3332, likely unintentional?

EmmettBicker replied:
omg I'm sorry you're right that's just a copy and paste issue. I'll try and be more mindful

janeyx99 reviewed (approved):
approval contingent on green CI + responding to my final comments. overall it is good

janeyx99 commented:
If the test is meant for only CUDA/CPU, you may need to add a decorator so it only runs for those (currently XLA is failing). There's this decorator onlyNativeDeviceTypes, but even in PyTorch we have MPS...

emmett-bicker committed:
Update test/test_torch.py

Co-authored-by: Jane (Yuan) Xu <31798555+janeyx99@users.noreply.github.com>

emmett-bicker committed:
simplified cuda errors test a lot

EmmettBicker commented:
I applied responses to all the comments!

EmmettBicker commented:
Oh nevermind I still need to change a variable name, one moment

emmett-bicker committed:
Changed variable name

EmmettBicker commented:
Now it should all be in

janeyx99 commented:
@pytorchbot merge

EmmettBicker commented:
Woohoo!

janeyx99 commented on test/test_torch.py:
```diff
@@ -3315,6 +3319,21 @@ def rand_tensor(size, dtype, device):
             self.assertTrue(not (out.isnan() or out.isinf()))
 
+    @onlyCUDA
+    def test_addcmul_cuda_errors_with_cpu_scalars(self, device):
```
lmao i see u gave up on the cpu part, which is fine 😛
